controller/worker/domain_migration/domain_migration.go

```diff
@@ -0,0 +1,358 @@
+
+	m := &migration{
+		db:     c.db,
+		client: c.client,
+		logger: log,
+		dm:     dm,
+	}
+
+	if err := m.db.QueryRow("SELECT old_domain, domain, old_tls_cert, tls_cert, created_at, finished_at FROM domain_migrations WHERE migration_id = $1", dm.ID).Scan(&dm.OldDomain, &dm.Domain, &dm.OldTLSCert, &dm.TLSCert, &dm.CreatedAt, &dm.FinishedAt); err != nil {
+		log.Error("error fetching postgres record", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	if dm.FinishedAt != nil {
+		// Already done
+		return nil
+	}
+
+	return m.Run()
+}
+
+func (m *migration) Run() error {
+	log := m.logger
+	dm := m.dm
+
+	// Generate TLS Cert if not already present
+	var cert *tlscert.Cert
+	if dm.TLSCert == nil {
+		var err error
+		cert, err = m.generateTLSCert()
+		if err != nil {
+			log.Error("error generating TLS cert", "err", err)
+			m.createEvent(err)
+			return err
+		}
+		dm.TLSCert = cert
+	}
+
+	if err := m.maybeDeployController(); err != nil {
+		log.Error("error deploying controller", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	if err := m.maybeDeployRouter(); err != nil {
+		log.Error("error deploying router", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	if err := m.maybeDeployDashboard(); err != nil {
+		log.Error("error deploying dashboard", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	if err := m.createMissingRoutes(); err != nil {
+		log.Error("error creating missing routes", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	if err := m.db.QueryRow("UPDATE domain_migrations SET finished_at = now() RETURNING finished_at").Scan(&dm.FinishedAt); err != nil {
+		log.Error("error setting finished_at", "err", err)
+		m.createEvent(err)
+		return err
+	}
+
+	log.Info("domain migration complete")
+
+	m.createEvent(nil)
+
+	return nil
+}
+
+func (m *migration) generateTLSCert() (*tlscert.Cert, error) {
+	hosts := []string{
+		m.dm.Domain,
+		fmt.Sprintf("*.%s", m.dm.Domain),
+	}
+	cert, err := tlscert.Generate(hosts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := m.db.Query("UPDATE domain_migrations SET tls_cert = $1 WHERE migration_id = $2", cert, m.dm.ID); err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func dupRelease(release *ct.Release) *ct.Release {
+	return &ct.Release{
+		ArtifactID: release.ArtifactID,
+		Env:        release.Env,
+		Meta:       release.Meta,
+		Processes:  release.Processes,
+	}
+}
+
+func (m *migration) waitForDeployment(app string) error {
+	list, err := m.client.DeploymentList(app)
+	if err != nil || len(list) == 0 {
+		return err
+	}
+	d := list[0]
+	if d.Status != "pending" && d.Status != "running" {
+		return nil
+	}
+
+	events := make(chan *ct.Event)
+	stream, err := m.client.StreamEvents(controller.StreamEventsOptions{
+		AppID:       d.AppID,
+		ObjectID:    d.ID,
+		ObjectTypes: []ct.EventType{ct.EventTypeDeployment},
+	}, events)
+	if err != nil {
+		return err
+	}
+	defer stream.Close()
+
+	log := m.logger.New("app", app, "deployment", d.ID)
```

Comment:
`app` should be something else as there is already a key with the same name. `deploy_app` maybe?